Blog on PSP deprecation (#26279)
kubernetes/website: content/en/blog/_posts/2021-01-28-PSP-deprecation.md

---
layout: blog

> title: 'What deprecation of Pod Security Policies means'

Reviewer: Comment from SIG Docs weekly meeting:
we recommend having “replace” or “replacement” in the article title, rather than “deprecate” or “deprecation”

Author: One thing we struggled with here is that SIG-Auth didn't have a specific suggestion for a replacement technology. (At least from what reviews we got from them.)  So "replacement" wouldn't be quite correct for what we're actually talking about here? We tried to arrange this post to explain that the deprecation process doesn't mean it's going away immediately, and that there are other options, but we don't want to recommend a specific one. Our suggestion really is to take a look at external options and choose the right one on a case-by-case basis.

Author: A new commit has been pushed to suggest a new title. Thank you for the feedback and please do check it out!

date: 2021-01-28
slug: psp-deprecation
---

**Authors:** Matt Broberg (Red Hat), Kaslin Fields (Google), Chris Short (Red Hat)

> Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This bears the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/). For more details, keep reading.

Reviewer: ```suggestion
Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This bears the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](/docs/reference/access-authn-authz/admission-controllers/). For more details, keep reading.
```

## What do Pod Security Policies do now?

> > A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritypolicy-v1beta1-policy) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.

Reviewer: ```suggestion
> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.
```

Reviewer: Take care when hyperlinking to https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritypolicy-v1beta1-policy as this will _not_ stay current after the v1.21 release. We are switching to a new API reference (that will have stable URIs), see suggestion above.

> Doc: [https://kubernetes.io/docs/concepts/policy/pod-security-policy/](https://kubernetes.io/docs/concepts/policy/pod-security-policy/)

Reviewer: ```suggestion
The [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/)
documentation explains this in more detail.
```
?

{{< youtube "SFtHRmPuhEw?start=953" >}} 

In short, PSPs are part of the kubelet and API server. The tight coupling to these parts of Kubernetes makes PSPs cumbersome both for developers and for operators. Using PSPs is often unintuitive, and setting them up too often leads to unintended results. The initial design of PSPs did not foresee many of the use cases they’ve since tried to solve.

To quote the above video: "Various limitations and structural problems have prevented the PSP API from GA." Having objects in Kubernetes never reaching GA (“permabeta”) is a problem for administrators and organizations, as a non-GA feature is not considered to be ready for widespread use. Given the structural problems with the PSPs as they exist today, there is no real path for them to become GA. Per the deprecation issue, “the future of restricting pod security settings does not lie in PSP because compatibility restrictions will prevent the kinds of changes that are required.” As such, PSPs will be deprecated in Kubernetes 1.21, though they will not be fully removed until a future version.

## What will replace them? 

> Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/) page in the Kubernetes Documentation aims “to detail recommended Pod security profiles, decoupled from any specific instantiation.”  This page is a useful reference for common pod security configurations, which can be implemented by native or external pod security enforcement tools.

Reviewer: ```suggestion
Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](/pod-security-standards/) page in the Kubernetes documentation aims “to detail recommended Pod security profiles, decoupled from any specific instantiation.”  This page is a useful reference for common pod security configurations, which can be implemented by native or external pod security enforcement tools.
```

> While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Controller Webhook in Kubernetes, but there are a variety of available options. The blog post, [“A Guide to Kubernetes Admission Controllers”](https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) from March 2019 gives a great overview of what admission controller webhooks can do.

Reviewer: ```suggestion
While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Controller Webhook in Kubernetes, but there are a variety of available options. The blog post, [“A Guide to Kubernetes Admission Controllers”](/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) from March 2019 gives a great overview of what admission controller webhooks can do.
```

While the exact capabilities and use of policy enforcement tools will vary, the community is engaged in working toward better and better solutions for enforcing security policies across your Kubernetes pods.

## Wrapping it up

Deprecation notices draw our attention to the wide range of work going on in the Kubernetes ecosystem that will change the way we operate Kubernetes clusters in the future. This work is taken quite seriously and goes through a detailed exploration process by related SIG leads to understand the implications.

> The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.

Reviewer: I worry about this message. My concern is that people who skip to the end of the article will infer that the overall message is:
1. PodSecurityPolicy is going
2. There isn't yet a viable alternative

and that this might not go down well with all folks.

Author: Thanks! I'll work on reworking this to be more positive and more of a summary. (Whereas it's currently more conclusion than summary.)

---
layout: blog
title: 'What’s Up with Pod Security Policies?'
date: 2021-02-04
slug: psp-update-2021
---

> **Authors:** Contributor Comms Team

Author: Added.

> **Authors:** Contributor Comms Team

Author: Does this need two new lines though?

> **Authors:** Matt Broberg (RedHat), Kaslin Fields (Google), Chris Short (RedHat)

Reviewer 2: ```suggestion
**Authors:** Matt Broberg (Red Hat), Kaslin Fields (Google), Chris Short (Red Hat)
```

> **Authors:** Matt Broberg (RedHat), Kaslin Fields (Google), Chris Short (RedHat)

Reviewer 2: I have a suggested edit in there for that above.

> 

Reviewer 2: ```suggestion
_The following is a brief overview of upcoming changes to Kubernetes. Bookmark and [keep an eye on our release site](https://www.kubernetes.dev/resources/release/) to see the latest news and other important updates._
```

Reviewer 2: ^ This is intended to reinforce the scope from the beginning of the article. If there's a better place to link them, let's do it here. cc @justaugustus @divya-mohan0209

Reviewer 3: @mbbroberg : I agree with this reinforcement of scope at the very start of the blog. It makes total sense to lay it out & convey that this is not a technical deep-dive and shouldn't be treated as one.

> Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This bears the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](/docs/reference/access-authn-authz/admission-controllers/). For more details, keep reading.

Author: Changed to "prompts"

## What do Pod Security Policies do now?

> > A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.

Reviewer: Also, tiny tweak:
```suggestion
> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.
```

Reviewer: Could make that italic (wrap it in underscores) if you like.

Doc: [https://kubernetes.io/docs/concepts/policy/pod-security-policy/](https://kubernetes.io/docs/concepts/policy/pod-security-policy/)

## Why are Pod Security Policies being deprecated? 

To watch an explanation of PSP removal, you can check out this session by SIG-Auth at KubeCon NA 2019:

> [https://youtu.be/SFtHRmPuhEw?t=953](https://youtu.be/SFtHRmPuhEw?t=953) 

Reviewer 2: Could we embed the video? Need to defer to a blog team member on whether this will work. 

```suggestion
<iframe width="560" height="315" src="https://www.youtube.com/embed/SFtHRmPuhEw?start=953" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

Reviewer 2: Ignore this -- we should use the Hugo short code instead of iframes.

> {{< youtube "SFtHRmPuhEw?start=953" >}} 

Reviewer: Looks massive on my (2160p) screen
```suggestion
{{< youtube "SFtHRmPuhEw?start=953" youtube-quote-sm >}}
```
perhaps?

While the exact capabilities and use of policy enforcement tools will vary, the community is engaged in working toward better and better solutions for enforcing security policies across your Kubernetes pods.

## Wrapping it up

Deprecation notices draw our attention to the wide range of work going on in the Kubernetes ecosystem that will change the way we operate Kubernetes clusters in the future. This work is taken quite seriously and goes through a detailed exploration process by related SIG leads to understand the implications.

> The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.

Reviewer 4: It's slotted to be removed in 1.25 but, that's far enough out for it to slip or move. We didn't necessarily want to lock the community into a hard and fast date/release.

> The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.

Reviewer 5: +1 for putting a concrete end-release instead of a vague "future version". The removal release is explicitly 1.25, and is surfaced in release notes and in warnings returned by the API server.

> The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.

Author: A new commit has been pushed to address this.

> The removal of PSP will pave the way for adoption of more suitable and extensible security policy enforcement options.

Reviewer 2: I get stuck on `suitable` -- on knowing what that means. It might flow better focused on extensibility. 

```suggestion
The removal of PSP will pave the way for adoption of more extensible security policy enforcement options.
```

Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This prompts the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](/docs/reference/access-authn-authz/admission-controllers/). For more details, keep reading.

## What do Pod Security Policies do now?

> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.

The [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/)
documentation explains this in more detail.

## Why are Pod Security Policies being deprecated? 

> To watch an explanation of PSP removal, you can check out this session by SIG-Auth at KubeCon NA 2019:

Reviewer 6: nit: "To watch an explanation of PSP removal ..."? Possibly reword to  something like:
To [better] understand the development of PSP ... and why PSP is being deprecated OR
To understand why PSP ...

Since the video is from 2019, I am guessing that your intent is to provide some historical perspective about the upcoming changes. This section describes the basic rationale for removing or deprecating PSP (without going into much technical detail). The video is backing up your claim and reasoning for the taking these steps to change.

{{< youtube "SFtHRmPuhEw?start=953" youtube-quote-sm >}}

> In short, PSPs are part of the kubelet and API server. The tight coupling to these parts of Kubernetes makes PSPs cumbersome both for developers and for operators. Using PSPs is often unintuitive, and setting them up too often leads to unintended results. The initial design of PSPs did not foresee many of the use cases they’ve since tried to solve.

Reviewer 7: I don't actually think there's a problem with the tight coupling to the API server. Both of the replacement designs we're considering are still built-in. [This document](https://docs.google.com/document/d/1VKqjUlpU888OYtIrBwidL43FOLhbmOD5tesYwmjzO4E/edit#heading=h.3qhd5uuytbqs) has a summary of what I see as the top issues with PSP. How about this as an alternative?
```suggestion
In short, there are 3 main challenges with PSPs today:
1. A flawed policy authorization model - Policies can be bound either to the pod's service account, or the requester. This dual binding approach weakens security, and often leads to confusion and breakage.
2. Rollout challenges - PSP is enabled or disabled for the entire cluster, lacks a non-enforcing or dryrun mode, and fails closed. This combination makes it very difficult to enable PSP by default, or rollout to an existing cluster.
3. Inconsistent API - PSP has grown organically over many releases but doesn't have a clear scope of what should be included. Backwards compatibility issues has lead to inconsistencies in the default behavior of several fields, and mutations can have unexpected side effects.
```

To quote the above video: "Various limitations and structural problems have prevented the PSP API from GA." Having objects in Kubernetes never reaching GA (“permabeta”) is a problem for administrators and organizations, as a non-GA feature is not considered to be ready for widespread use. Given the structural problems with the PSPs as they exist today, there is no real path for them to become GA. Per the deprecation issue, “the future of restricting pod security settings does not lie in PSP because compatibility restrictions will prevent the kinds of changes that are required.” As such, PSPs will be deprecated in Kubernetes 1.21, though they will not be fully removed until version 1.25.

> ## What will replace them? 

Reviewer 6: nit: replace "them" with Pod Security Policies?
nit: "While the scope of PSPs is beyond anything that will replace it". This paragraph is somewhat vague. Will you indicate that more communication/documentation about ACWs and security policies is coming.

> ## What will replace them? 

Reviewer 7: When is this targeted for publication? We are still hoping to have agreed upon a replacement design prior to v1.21. Even if we haven't selected a replacement design, I think it's worth mentioning that there is a sig-auth breakout group working towards a replacement.

Reviewer: I don't think a publication date is set yet.

Reviewer 4: This should've gone out weeks ago when people started asking me.

Reviewer 8: Personally, I'm glad this post didn't go out weeks ago, because SIGs Auth and Security have been making good progress on a solution that was not reflected in this post. At this time, we are planning on closing this PR, rewriting a post to more accurately reflect current events, and putting in a new one. We feel that a clearer post with more definitive solutions will provide better value for users overall than sooner publication of a post without those things would have. Thanks! See you in that PR :)

Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](/docs/concepts/security/pod-security-standards/) page in the Kubernetes documentation aims “to detail recommended Pod security profiles, decoupled from any specific instantiation.”  This page is a useful reference for common pod security configurations, which can be implemented by native or external pod security enforcement tools.

> While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Controller Webhook in Kubernetes, but there are a variety of available options. The blog post, [“A Guide to Kubernetes Admission Controllers”](/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) from March 2019 gives a great overview of what admission controller webhooks can do.

Reviewer 7: > While the scope of PSPs is beyond anything that will replace it

I'm not sure what this means. Do you mean that none of the replacements we're considering cover the full scope of what PSPs do?

> While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Controller Webhook in Kubernetes, but there are a variety of available options. The blog post, [“A Guide to Kubernetes Admission Controllers”](/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) from March 2019 gives a great overview of what admission controller webhooks can do.

Reviewer 7: > Pod Security Policies are one type of Admission Controller Webhook in Kubernetes

PSPs are a built in controller, not a webhook

While the exact capabilities and use of policy enforcement tools will vary, the community is engaged in working toward better and better solutions for enforcing security policies across your Kubernetes pods.

## Wrapping it up

Pod Security Policies were an early attempt to support users' needs with regard to securing pods. Since their creation, they have become unwieldy within the larger context of Kubernetes, making them both difficult to maintain and difficult for users to use.

> 

Reviewer 6: nit; Possibly reword, "... for users to use"? Something like:
"making them difficult to configure, manage, and maintain."
